### PR TITLE
DOC Document changes to template layer

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/11322

### CODE REVIEW

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 1]
```diff
@@ -141,6 +141,12 @@ $SilverStripeNavigator
 $Content
 ```
 
+### Case sensitivity
+
+Just like in PHP, method names are case insensitive, but property names are case sensitive.
+
+Using `$MyValue` in a template will successfully call a `myValue()` method, or a `getMyvalue()` method even though the case doesn't match what you used in the template. But it *will not* match a `myValue` property or database field because the case doesn't match.
```
Comment (emteknetnz) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md]:
```suggestion
Using `$MyValue` in a template will successfully call a `myValue()` method, or a `getMyValue()` method even though the case doesn't match what you used in the template. But it *will not* match a `myValue` property or database field because the case doesn't match.
```

File: en/08_Changelogs/6.0.0.md [hunk 2]
```diff
@@ -459,6 +459,62 @@ Note that the `SilverStripe\View\ViewableData` class has been renamed to [`Silve
 
 See [many renamed classes](#renamed-classes) for more information about this change.
 
+#### Improved separation between the view and model layers {#view-layer-separation}
+
+Historically the `ModelData` class did double-duty as being the base class for most models as well as being the presumed class wrapping data for the template layer. Part of this included methods like `XML_val()` being called on any object in the template layer, despite being methods very specifically implemented on `ModelData`.
```
Comment (emteknetnz) [inline en/08_Changelogs/6.0.0.md:466]: Historically it's been called `ViewableData` ? Maybe mention that ModelData used to be called ViewableData

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 3]
```diff
@@ -69,32 +68,33 @@ This will result in trying to fetch a `Title` property from the data being displ
 Variables can be chained together, and include arguments.
 
 ```ss
-$Foo(param)
+$Foo("param")
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:71]: mostly-unrelated update to promote best practice

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 4]
```diff
@@ -69,32 +68,33 @@ This will result in trying to fetch a `Title` property from the data being displ
...
 $Foo.Bar
 ```
 
 These variables will call a method / field on the object and insert the returned value as a string into the template.
 
+Arguments don't have to be literals - you can pass template variables as arguments as well.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:77]: This is new and was mostly just a natural consequence of how we're wrapping data now. Very useful though, I'm already gonna use it in the CMSMain refactor.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 5]
```diff
@@ -69,32 +68,33 @@ This will result in trying to fetch a `Title` property from the data being displ
...
+
 > [!WARNING]
-> If you wish to pass arguments to getter functions, you must use the full method name. e.g. `$Thing` will try to access `Thing` as a property, which will ultimately result in `getThing()` being called with no arguments To pass arguments you must use `$getThing('param')`.
->
-> Also, arguments must be literals, and cannot be other template variables (`$getThing($variable)` will pass the literal string `'$variable'` to the `getThing()` method).
+> Arguments cannot be literal arrays (e.g `$myMethod(["val1", "val2"])`). The behaviour for this is currently undefined.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:80]: This was mostly saying "you can't pass template variables as arguments" which as noted above is no longer the case. It's just array literals you can't do now.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 6]
```diff
@@ -69,32 +68,33 @@ This will result in trying to fetch a `Title` property from the data being displ
...
 
 - `$Foo("param")` will call `$obj->Foo("param")`
 - `$Foo.Bar` will call `$obj->Foo()->Bar`
+- `$Foo($Bar)` will call `$obj->Bar` and then pass the result into `$obj->Foo($result)`
 
 > [!NOTE]
-> Arguments passed into methods can be any non-array literal type (not just strings), e.g:
+> Arguments passed into methods can be almost any type (not just strings), e.g:
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:87]: At first glance you might think to leave as-is cause we still can't do arrays - but it doesn't have to be a literal anymore which the old text said it does.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 7]
```diff
@@ -111,8 +111,7 @@ use SilverStripe\ORM\DataObject;
 class MyObject extends DataObject
 {
     // ...
-
-    public function UsersIpAddress()
+    public function usersIpAddress()
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:114]: Unrelated change to promote best practice. Also updated casing on the below.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md (hunk 1 above)
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md]: This context was missing. This is the way it has always worked, I believe.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 8]
```diff
@@ -564,39 +570,24 @@ refer directly to properties and methods of the [`Member`](api:SilverStripe\Secu
...
-    public function forTemplate(): string
-    {
-        // We can also render a template here using $this->renderWith()
-        return 'Page: ' . $this->Title;
-    }
-}
-```
+If you reference some object directly in a template, the `forTemplate()` method on that object will be called. This can be used to provide a default template for an object.
+If the `forTemplate()` method isn't implemented, the system will attempt to cast your object to a string.
+The default implementation of this method on `ModelData` renders the model using templates named after the class or its superclasses.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:575]: No need to show an example, since there is now a default implementation which will attempt to render templates by default.

File: en/02_Developer_Guides/01_Templates/01_Syntax.md [hunk 9]
```diff
@@ -564,39 +570,24 @@ refer directly to properties and methods of the [`Member`](api:SilverStripe\Secu
...
+<%-- calls forTemplate() on the first page in the list --%>
 $Pages->First
 ```
 
 You can also use the `$Me` variable, which outputs the current object in scope by calling `forTemplate()` on the object.
 This is especially helpful when you want to directly render items in a list you're looping over.
 
 > [!WARNING]
-> If the object does not have a `forTemplate()` method implemented, this will throw an error.
+> If the object does not have an appropriate template, implement the `forTemplate()` method, or implement [`__toString()`](https://www.php.net/manual/en/language.oop5.magic.php#object.tostring), this will throw an error.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/01_Syntax.md:586]: Objects don't have to be `ModelData` which is why this warning allows for the case where `forTemplate()` isn't implemented.

File: en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md [hunk 10]
```diff
@@ -194,5 +203,62 @@ class MyPageController extends PageController
 }
 ```
 
-> [!WARNING]
-> A common mistake is trying to loop over an array directly in a template - this won't work. You'll need to wrap the array in some `ModelData` instance as mentioned above.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md:198]: Not the case anymore!

File: en/02_Developer_Guides/01_Templates/05_Template_Inheritance.md [hunk 11]
```diff
@@ -103,21 +103,6 @@ to determine resolution priority. This search is based on the following three co
...
-
-For each path the loader will search in this order:
-
-- Loop through each theme which is configured.
-- If a theme is a set (declared with the `$` prefix, e.g. `$default`) it will perform a nested search within
-   that set.
-- When searching the `$default` set, all modules will be searched in the order declared via the `module_priority`
-   config, interpolating keys `$project` and `$other_modules` as necessary. By default, your project's templates should
-   be checked before any vendor modules.
-- When the first template is found, it will be immediately returned, and will not continue to search.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/05_Template_Inheritance.md:119]: That class no longer has anything to do with templates.

File: en/02_Developer_Guides/01_Templates/07_Caching.md [hunk 12]
```diff
@@ -8,8 +8,8 @@ icon: rocket
 
 ## Object caching
 
-All functions that provide data to templates must have no side effects, as the value is cached after first access. For
-example, this controller method will not behave as you might imagine.
+Any values accessed from a template on a `ModelData` object are cached after first access. Because of this, methods that provide data to templates should ideally have no side effects. For
+example, this `getCounter()` method will not behave as you might imagine when invoked from a template.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/07_Caching.md:12]: This level of caching is specific to `ModelData`. Always has been.

File: en/02_Developer_Guides/01_Templates/07_Caching.md [hunk 13]
```diff
@@ -51,3 +50,9 @@ Example:
     $CacheableContent
 <% end_cached %>
 ```
+
+## Template caching
+
+Every time a raw template file is processed, some PHP code is generated from it which is then executed to produce the final rendered result.
+
+That PHP code is stored in the filesystem so that the raw template file doesn't need to be processed again until either the cache is flushed or the raw template file is updated.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/07_Caching.md]: This context was missing before.

File: en/02_Developer_Guides/01_Templates/09_Casting.md [hunk 14]
```diff
@@ -6,13 +6,17 @@ icon: code
 
 # Formatting, casting, and escaping variable content
 
-All objects that are being rendered in a template should be a [ModelData](api:SilverStripe\Model\ModelData) instance such as `DataObject`,
-`DBField` or `Controller`. From these objects, the template can include any method from the object in [scope](syntax#scope).
+Classes that subclass [`ModelData`](api:SilverStripe\Model\ModelData) (such as `DataObject`,
+`DBField` or `Controller`) can define how their values should be cast or escaped when used in a template.
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/01_Templates/09_Casting.md:10]: You can pass basically anything into templates, doesn't have to be `ModelData` - but the ability to define casting like this _is_ unique to `ModelData` and its subclasses.

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/04_Preview.md [hunk 15]
```diff
@@ -272,27 +272,6 @@ list of items and indirectly calling `forTemplate` using the [`$Me` template var
...
-    // ...
-
-    public function forTemplate(): string
-    {
-        // If the template for this DataObject is not an "Include" template, use the appropriate type here
-        // e.g. "Layout".
-        return $this->renderWith(['type' => 'Includes', self::class]);
-    }
-}
-```
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/04_Preview.md:294]: Remove unnecessary example now that the default implementation in `ModelData` does this for you.

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/04_Preview.md [hunk 16]
```diff
@@ -405,7 +383,7 @@ class MyAdmin extends ModelAdmin
 >         // Render the preview content
 >         $preview = $obj->forTemplate();
 >         // Wrap preview in proper html, body, etc so Requirements are used
->         $preview = SSViewer::create('PreviewBase')->process(ArrayData::create(['Preview' => $preview]));
+>         $preview = SSViewer::create('PreviewBase')->process(['Preview' => $preview]);
```
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/04_Preview.md:386]: No need to wrap them anymore. I'd argue it's even best practice to not wrap them.

File: en/08_Changelogs/6.0.0.md [hunk 17]
```diff
@@ -526,12 +582,12 @@ You can change the level of entropy required by passing one of the valid [minSco
 
 ### Other new features
 
-- Native indexed PHP arrays can now be passed into templates and iterated over with `<% loop $MyArray %>`. Under the hood they are wrapped in [`ArrayList`](api:SilverStripe\Model\List\ArrayList), so you can get the count using `$Count` and use `<% if $ArrayList %>` as a shortcut for `<% if $ArrayList.Count %>`. Other functionality from `ArrayList` such as filtering and sorting cannot be used on arrays since they don't have keys to filter or sort against.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md:529]: Moved this into the template changes section

File: en/08_Changelogs/6.0.0.md [hunk 18]
```diff
@@ -526,12 +582,12 @@ You can change the level of entropy required by passing one of the valid [minSco
...
 - A new [`DataObject::getCMSEditLink()`](api:SilverStripe\ORM\DataObject::getCMSEditLink()) method has been added, which returns `null` by default. This provides more consistency for that method which has previously been inconsistently applied to various subclasses of `DataObject`. See [managing records](/developer_guides/model/managing_records/#getting-an-edit-link) for more details about providing sane values for this method in your own subclasses.
 - The `CMSEditLink()` method on many `DataObject` subclasses has been renamed to `getCMSEditLink()`.
 - The [`UrlField`](api:SilverStripe\Forms\UrlField) class has some new API for setting which protocols are allowed for valid URLs.
 - The [`EmailField`](api:SilverStripe\Forms\EmailField) class now uses `symfony/validator` to handle its validation logic, where previously this was validated with a custom regex.
+- [`ArrayData`](api:SilverStripe\Model\ArrayData) can now be serialised using [`json_encode()`](https://www.php.net/manual/en/function.json-encode.php).
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md:590]: This isn't directly related to templates so adding here instead.

File: en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md [hunk 19]
```diff
@@ -6,6 +6,9 @@ icon: code
 
 # Rendering data to a template
 
+> [!NOTE]
+> The template syntax, file extensions, and specifics about which tempaltes are chosen from a set as described on this page are specific to the default [`SSTemplateEngine`](api:SilverStripe\View\SSTemplateEngine) - but many of the concepts here (especially the PHP code) should work with any template engine you choose to use.
```
Comment (emteknetnz) [inline en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md]:
```suggestion
> The template syntax, file extensions, and specifics about which templates are chosen from a set as described on this page are specific to the default [`SSTemplateEngine`](api:SilverStripe\View\SSTemplateEngine) - but many of the concepts here (especially the PHP code) should work with any template engine you choose to use.
```

File: en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md [hunk 20]
```diff
@@ -194,5 +203,62 @@ class MyPageController extends PageController
 }
 ```
 
-> [!WARNING]
-> A common mistake is trying to loop over an array directly in a template - this won't work. You'll need to wrap the array in some `ModelData` instance as mentioned above.
+Or wrap the data in a `ModelData` subclass:
```
Comment (emteknetnz) [inline en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md]:
```suggestion
Or wrap the data in a `ModelData` subclass such as `ArrayList`:
```

File: en/02_Developer_Guides/01_Templates/07_Caching.md (hunk 13 above)
Comment (emteknetnz) [inline en/02_Developer_Guides/01_Templates/07_Caching.md]:
```suggestion
That PHP code is cached in the filesystem so that the raw template file doesn't need to be processed again until either the cache is flushed or the raw template file is updated.
```

Comment (emteknetnz) [reply en/02_Developer_Guides/01_Templates/07_Caching.md]: Just ties it together a little better

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: en/02_Developer_Guides/01_Templates/01_Syntax.md (hunk 1 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/01_Templates/01_Syntax.md]: Done (found an unrelated method with the same bad casing in php coding conventions which I also updated)

File: en/08_Changelogs/6.0.0.md (hunk 2 above)
Comment (ChloeSartorelli) [reply en/08_Changelogs/6.0.0.md:466]: Directly above this it says
> Note that the `SilverStripe\View\ViewableData` class has been renamed to `SilverStripe\Model\ModelData`. We will refer to it as `ModelData` in the rest of these change logs.

File: en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md (hunk 19 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md]: Done

File: en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md (hunk 20 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/01_Templates/04_Rendering_Templates.md]: Done, but said `ArrayData` since that's what the example shows.

File: en/02_Developer_Guides/01_Templates/07_Caching.md (hunk 13 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/01_Templates/07_Caching.md]: Done

**Review by emteknetnz — APPROVED** [review]
